src/ints/xms.cpp

```diff
@@ -16,478 +17,843 @@
+
 	// Block has to be unlocked
-	if (xms_handles[handle].locked>0) return XMS_BLOCK_LOCKED;
-	Bitu pages=newSize/4 + ((newSize & 3) ? 1 : 0);
-	if (MEM_ReAllocatePages(xms_handles[handle].mem,pages,true)) {
-		xms_handles[handle].size = newSize;
-		return 0;
-	} else return XMS_OUT_OF_SPACE;
+
+	if (xms.handles[handle].lock_count > 0) {
+		return Result::XMS_BLOCK_LOCKED;
+	}
+
+	constexpr bool sequence = true;
+	if (MEM_ReAllocatePages(xms.handles[handle].mem_handle,
+	                        get_num_pages(new_size_kb),
+	                        sequence)) {
+		xms.handles[handle].size_kb = new_size_kb;
+		return Result::OK;
+	}
+
+	return Result::XMS_OUT_OF_SPACE;
 }
 
-static bool multiplex_xms(void) {
+static bool xms_multiplex()
+{
 	switch (reg_ax) {
-	case 0x4300:					/* XMS installed check */
-			reg_al=0x80;
-			return true;
-	case 0x4310:					/* XMS handler seg:offset */
-			SegSet16(es,RealSeg(xms_callback));
-			reg_bx=RealOff(xms_callback);
-			return true;			
+	case 0x4300: // XMS installed check
+		reg_al = 0x80;
+		return true;
+	case 0x4310: // XMS handler seg:offset
+		SegSet16(es, RealSeg(xms.callback));
+		reg_bx = RealOff(xms.callback);
+		return true;
 	}
-	return false;
 
+	return false;
 }
 
-inline void SET_RESULT(Bitu res,bool touch_bl_on_succes=true) {
-	if(touch_bl_on_succes || res) reg_bl = (uint8_t)res;
-	reg_ax = (res==0);
-}
+// ***************************************************************************
+// Main XMS API handler
+// ***************************************************************************
+
+static Bitu XMS_Handler()
+{
+	assert(xms.is_available);
+
+	Result result = Result::OK;
+
+	auto set_return_value = [](const Result result) {
+		reg_bl = static_cast<uint8_t>(result);
+		reg_ax = (result == Result::OK) ? 1 : 0;
+	};
+
+	auto set_return_value_bl_only_fail = [](const Result result) {
+		if (result != Result::OK) {
+			reg_bl = static_cast<uint8_t>(result);
+		}
+		reg_ax = (result == Result::OK) ? 1 : 0;
+	};
 
-Bitu XMS_Handler(void) {
-//	LOG(LOG_MISC,LOG_ERROR)("XMS: CALL %02X",reg_ah);
 	switch (reg_ah) {
-	case XMS_GET_VERSION:										/* 00 */
-		reg_ax=XMS_VERSION;
-		reg_bx=XMS_DRIVER_VERSION;
-		reg_dx=0;	/* No we don't have HMA */
+	case 0x00: // Get XMS Version Number
+		reg_ax = XMS_VERSION;
+		reg_bx = XMS_DRIVER_VERSION;
+		reg_dx = hma.is_available ? 1 : 0;
 		break;
-	case XMS_ALLOCATE_HIGH_MEMORY:								/* 01 */
-		reg_ax=0;
-		reg_bl=HIGH_MEMORY_NOT_EXIST;
+	case 0x01: // Request High Memory Area
+		if (!hma.is_available) {
+			set_return_value(Result::HMA_NOT_EXIST);
+			break;
+		}
+		if (hma.app_has_control || hma.dos_has_control) {
+			// HMA already controlled by application or DOS
+			set_return_value(Result::HMA_IN_USE);
+		} else if (reg_dx < hma.min_alloc_size) {
+			// Request for a block not big enough
+			set_return_value(Result::HMA_NOT_BIG_ENOUGH);
+		} else {
+			reg_ax = 1; // HMA allocated succesfully
+			LOG_INFO("XMS: HMA allocated by application/TSR");
```

Comment:
I don't think regular users care about such logs, should be DEBUG level.

Reply:
OK, changed.